File: python/tests/test_conformance.py
"""Conformance tests against shared testdata."""

from __future__ import annotations

import json
from pathlib import Path

import pytest

from spanvalue import (
    FORMAT_OPTION_MORE_VERBOSE,
    FORMAT_OPTION_NORMAL,
    FORMAT_OPTION_SIMPLE,
    FORMAT_OPTION_SIMPLEST,
    FORMAT_OPTION_VERBOSE,
    LiteralQuoteConfig,
    PreferredQuote,
    QuoteStrategy,
    format_type,
    format_type_verbose_annotation_omit,
    format_type_verbose_annotation_primary,
    format_value,
    literal_format_config,
    simple_format_config,
    spanner_cli_format_config,
)

CONFORMANCE_PATH = Path(__file__).resolve().parents[2] / "testdata" / "conformance.json"

TYPE_PRESETS = {
    "simplest": FORMAT_OPTION_SIMPLEST,
    "simple": FORMAT_OPTION_SIMPLE,
    "normal": FORMAT_OPTION_NORMAL,
    "verbose": FORMAT_OPTION_VERBOSE,
    "more_verbose": FORMAT_OPTION_MORE_VERBOSE,
}

QUOTE_POLICIES = {
    "legacy_double": LiteralQuoteConfig(QuoteStrategy.LEGACY, PreferredQuote.DOUBLE),
    "legacy_single": LiteralQuoteConfig(QuoteStrategy.LEGACY, PreferredQuote.SINGLE),
    "always_double": LiteralQuoteConfig(QuoteStrategy.ALWAYS, PreferredQuote.DOUBLE),
    "always_single": LiteralQuoteConfig(QuoteStrategy.ALWAYS, PreferredQuote.SINGLE),
    "min_escape_double": LiteralQuoteConfig(QuoteStrategy.MIN_ESCAPE, PreferredQuote.DOUBLE),
    "min_escape_single": LiteralQuoteConfig(QuoteStrategy.MIN_ESCAPE, PreferredQuote.SINGLE),
}


def _load_conformance() -> dict:
    with CONFORMANCE_PATH.open(encoding="utf-8") as f:
        return json.load(f)


def _format_type_preset(name: str, typ: dict) -> str:
    if name == "verbose_annotation_omit":
        return format_type_verbose_annotation_omit(typ)
    if name == "verbose_annotation_primary":
        return format_type_verbose_annotation_primary(typ)
    return format_type(typ, TYPE_PRESETS[name])


@pytest.fixture(scope="module")
def conformance() -> dict:
    return _load_conformance()


@pytest.mark.parametrize("preset", list(TYPE_PRESETS) + ["verbose_annotation_omit", "verbose_annotation_primary"])
def test_type_cases(conformance: dict, preset: str) -> None:
    for case in conformance["type_cases"]:
        got = _format_type_preset(preset, case["type"])
        want = case["expected"][preset]
        assert got == want, f"type case {case['name']!r} preset {preset!r}: got {got!r} want {want!r}"


@pytest.mark.parametrize("preset", ["simple", "literal", "spanner_cli"])
def test_value_cases(conformance: dict, preset: str) -> None:
    if preset == "simple":
        config = simple_format_config()
    elif preset == "literal":
        config = literal_format_config()
    else:
        config = spanner_cli_format_config()

    for case in conformance["value_cases"]:
        got = format_value(case["type"], case["value"], config)
        want = case["expected"][preset]
        assert got == want, f"value case {case['name']!r} preset {preset!r}: got {got!r} want {want!r}"


@pytest.mark.parametrize("policy_name", list(QUOTE_POLICIES))
def test_value_literal_quotes(conformance: dict, policy_name: str) -> None:
    config = literal_format_config(quote=QUOTE_POLICIES[policy_name])
    for case in conformance["value_cases"]:
        got = format_value(case["type"], case["value"], config)
        want = case["expected"]["literal_quotes"][policy_name]
        assert got == want, f"value case {case['name']!r} quote {policy_name!r}: got {got!r} want {want!r}"

File: ruby/test/test_conformance.rb
# frozen_string_literal: true

require 'json'
require 'minitest/autorun'
require 'pathname'

require_relative '../lib/spanvalue'

class ConformanceTest < Minitest::Test
  CONFORMANCE_PATH = Pathname.new(__dir__).join('../../testdata/conformance.json').expand_path

  TYPE_PRESETS = {
    'simplest' => Spanvalue::FORMAT_OPTION_SIMPLEST,
    'simple' => Spanvalue::FORMAT_OPTION_SIMPLE,
    'normal' => Spanvalue::FORMAT_OPTION_NORMAL,
    'verbose' => Spanvalue::FORMAT_OPTION_VERBOSE,
    'more_verbose' => Spanvalue::FORMAT_OPTION_MORE_VERBOSE
  }.freeze

  QUOTE_POLICIES = {
    'legacy_double' => Spanvalue::LiteralQuoteConfig.new(
      strategy: Spanvalue::QuoteStrategy::LEGACY,
      preferred_quote: Spanvalue::PreferredQuote::DOUBLE
    ),
    'legacy_single' => Spanvalue::LiteralQuoteConfig.new(
      strategy: Spanvalue::QuoteStrategy::LEGACY,
      preferred_quote: Spanvalue::PreferredQuote::SINGLE
    ),
    'always_double' => Spanvalue::LiteralQuoteConfig.new(
      strategy: Spanvalue::QuoteStrategy::ALWAYS,
      preferred_quote: Spanvalue::PreferredQuote::DOUBLE
    ),
    'always_single' => Spanvalue::LiteralQuoteConfig.new(
      strategy: Spanvalue::QuoteStrategy::ALWAYS,
      preferred_quote: Spanvalue::PreferredQuote::SINGLE
    ),
    'min_escape_double' => Spanvalue::LiteralQuoteConfig.new(
      strategy: Spanvalue::QuoteStrategy::MIN_ESCAPE,
      preferred_quote: Spanvalue::PreferredQuote::DOUBLE
    ),
    'min_escape_single' => Spanvalue::LiteralQuoteConfig.new(
      strategy: Spanvalue::QuoteStrategy::MIN_ESCAPE,
      preferred_quote: Spanvalue::PreferredQuote::SINGLE
    )
  }.freeze

  def self.conformance
    @conformance ||= JSON.parse(File.read(CONFORMANCE_PATH, encoding: 'UTF-8'))
  end

  def format_type_preset(name, typ)
    case name
    when 'verbose_annotation_omit'
      Spanvalue.format_type_verbose_annotation_omit(typ)
    when 'verbose_annotation_primary'
      Spanvalue.format_type_verbose_annotation_primary(typ)
    else
      Spanvalue.format_type(typ, TYPE_PRESETS.fetch(name))
    end
  end

  TYPE_PRESETS.each_key do |preset|
    define_method("test_type_cases_#{preset}") do
      self.class.conformance.fetch('type_cases').each do |case_data|
        got = format_type_preset(preset, case_data['type'])
        want = case_data.fetch('expected').fetch(preset)
        assert_equal want, got, "type case #{case_data['name'].inspect} preset #{preset.inspect}"
      end
    end
  end

  %w[verbose_annotation_omit verbose_annotation_primary].each do |preset|
    define_method("test_type_cases_#{preset}") do
      self.class.conformance.fetch('type_cases').each do |case_data|
        got = format_type_preset(preset, case_data['type'])
        want = case_data.fetch('expected').fetch(preset)
        assert_equal want, got, "type case #{case_data['name'].inspect} preset #{preset.inspect}"
      end
    end
  end

  %w[simple literal spanner_cli].each do |preset|
    define_method("test_value_cases_#{preset}") do
      config = case preset
               when 'simple' then Spanvalue.simple_format_config
               when 'literal' then Spanvalue.literal_format_config
               else Spanvalue.spanner_cli_format_config
               end

      self.class.conformance.fetch('value_cases').each do |case_data|
        got = Spanvalue.format_value(case_data['type'], case_data['value'], config)
        want = case_data.fetch('expected').fetch(preset)
        assert_equal want, got, "value case #{case_data['name'].inspect} preset #{preset.inspect}"
      end
    end
  end

  QUOTE_POLICIES.each_key do |policy_name|
    define_method("test_value_literal_quotes_#{policy_name}") do
      config = Spanvalue.literal_format_config(QUOTE_POLICIES.fetch(policy_name))
      self.class.conformance.fetch('value_cases').each do |case_data|
        got = Spanvalue.format_value(case_data['type'], case_data['value'], config)
        want = case_data.fetch('expected').fetch('literal_quotes').fetch(policy_name)
        assert_equal want, got, "value case #{case_data['name'].inspect} quote #{policy_name.inspect}"
      end
    end
  end
end
